Add boilerplate unit test code to currently untested modules

hazzlim opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
In pursuit of increasing unit test coverage of MbedOS, we add the
boilerplate code for unit testing of AnalogIn.cpp and an empty test
source file. This serves two purposes - it allows us to report on
currently untested sources in code coverage data, and it makes it a
little easier for developers to write unit tests for these sources.

The 'empty' test file contains a main function that simply returns. This
is required to allow CMake's add_executable() to be used to pull in the
source file for the SUT, which ensures that this source file is built
and therefore instrumented to generate coverage data. The alternative
that was explored was to instead use Google Test's TEST() macro and
prefix the test name with 'DISABLED_' to skip it, but that resulted in
the test being reported as skipped, which was deemed undesireable for
these 'empty' tests.

Additionally, stubs for analogin_api.c functions are added to the Mbed-stubs-hal library, to allow AnalogIn.cpp to build & to be used by the future unit tests.
#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None.
### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None.
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@Patater @rwalton-arm @ARMmbed/mbed-os-core 
----------------------------------------------------------------------------------------------------------------

hazzlim committed:
Unit tests: Add boilerplate code for AnalogIn.cpp

In pursuit of increasing unit test coverage of Mbed OS, we add the
boilerplate code for unit testing of AnalogIn.cpp and an empty test
source file. This serves two purposes - it allows us to report on
currently untested sources in code coverage data, and it makes it a
little easier for developers to write unit tests for these sources.

The 'empty' test file contains a main function that simply returns. This
is required to allow CMake's add_executable() to be used to pull in the
source file for the SUT, which ensures that this source file is built
and therefore instrumented to generate coverage data. The alternative
that was explored was to instead use Google Test's TEST() macro and
prefix the test name with 'DISABLED_' to skip it, but that resulted in
the test being reported as skipped, which was deemed undesireable for
these 'empty' tests.

hazzlim committed:
Unit tests: Add stubs for analogin_api.c functions

Currently there are no stub implementations of the analogin_api.c
functions. As the AnalogIn class makes use of these functions, these
stub definitions are required in order to build AnalogIn.cpp and
generate .gcno files to be used to generate code coverage metrics.

rwalton-arm commented on drivers/tests/UNITTESTS/AnalogIn/test_analogin.cpp:
```diff
@@ -0,0 +1,21 @@
+ */
+
+// Replace this main function with Google Test tests, as described in the
+// documentation: https://os.mbed.com/docs/mbed-os/v6.12/debug-test/unit-testing.html
```
Should we link to the "latest" docs? Seems like this could get out of date if we pin it to the v6.12 docs. I think if we replace `/v6.12` with `/latest` it should redirect to the latest revision.

hazzlim replied:
Aha very good point, I will amend the url

hazzlim replied:
Done :)

rwalton-arm reviewed (commented):
LGTM, only one question/comment.

Patater commented on drivers/tests/UNITTESTS/AnalogIn/test_analogin.cpp:
```diff
@@ -0,0 +1,21 @@
+ */
+
+// Replace this main function with Google Test tests, as described in the Mbed OS
+// documentation: https://os.mbed.com/docs/mbed-os/latest/debug-test/unit-testing.html
```
An empty test file looks pretty weird to a developer browsing the code. We should comment in the source code (not just the commit message) that the file is present and empty in order to generate coverage metrics. Otherwise, a developer is very tempted to just delete this file, not knowing what purpose it is serving, as it looks like it serves no purpose.

hazzlim replied:
I have updated the comment to specify the purpose of it's existence!

Patater commented on hal/tests/UNITTESTS/doubles/analogin_api_stub.c:
```diff
@@ -0,0 +1,49 @@
+
+#if DEVICE_ANALOGIN
+
+void analogin_init_direct(analogin_t *obj, const PinMap *pinmap)
```
Just an observation; correct me if I'm wrong on anything. This is stubbing the HAL `analogin` interface, not the Mbed C++ `AnalogIn` class, so that we can build and link the real `AnalogIn` implementation and have the lack of testing of the Mbed C++ `AnalogIn` interface show up in unit test coverage metrics. Without at least an empty unit test that builds links that includes `AnalogIn`, we don't get any `.gcno` files generated (`AnalogIn.gcno`) and the `AnalogIn` class does not get included in any unit test coverage metrics.

hazzlim replied:
I have updated the commit message to reflect the reason we need to add these stubs.

hazzlim commented on drivers/tests/UNITTESTS/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) 2021 ARM Limited. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 add_subdirectory(doubles)
+add_subdirectory(AnalogIn)
```
For discussion:
Do we want to only build the empty tests when we are building with the COVERAGE option set? This saves us building all these untested files for no reason on a routine unit test build.

If so:
We could make inclusion of all of the empty test directories conditional
```
if(COVERAGE)
    add_subdirectory(AnalogIn)
    add_subdirectory(SomeOtherComponent)
endif()
```
Or put the conditional test in the CMakeLists.txt for the individual test file, which has the benefit of confining to that test directory the changes that need to be made when actually implementing the unit test.
```
if(COVERAGE)
    include(GoogleTest)
    ...
    gtest_discover_tests(${TEST_NAME} PROPERTIES LABELS "drivers")
endif()
```

Patater replied:
I think adding this makes it slightly more annoying to add the real unit tests later, compromising our goal of having this be useful not just for coverage metrics, but also as a decent starting point for adding the tests.

hazzlim replied:
Yes I'm inclined to agree, and streamlining/encouraging the adding of tests is the primary goal.

0xc0170 commented on drivers/tests/UNITTESTS/AnalogIn/test_analogin.cpp:
```diff
@@ -0,0 +1,30 @@
+ *      https://os.mbed.com/docs/mbed-os/latest/debug-test/unit-testing.html
+ *
+ */
+int main() {
```
`{` new line (I assume the astyle run in github action will catch this (wait, Travis run and it's green, I'll verify).

hazzlim replied:
Oops! Thanks for spotting, will amend

0xc0170 replied:
`^drivers/tests/UNITTESTS` is in codeignore. I wonder why we add almost all unittests into codeignore , I've missed this being added.

Patater replied:
We should add a commit to remove that folder from codeignore. We ought to be checking coding style in our unit tests.

LDong-Arm replied:
When we reworked the directory structure and moved files around last year, there were a number of style non-compliances in unit tests, and the decision was to ignore them and come back to this in the future. But fixing them all at once is super easy - just run astyle on all of them and they'll get reformatted, and we should've done this back then.

hazzlim replied:
I have ran astyle on everything and removed the directory from the codecheckignore.

0xc0170 replied:
nice!

hazzlim committed:
Style: Remove drivers unit test directory from codeignore

Unit tests for drivers are currently not checked for style compliance.
We should be checking coding style in unit tests, so we remove them from
the .codecheckignore file.

Patater commented on .codecheckignore:
```diff
@@ -35,4 +35,3 @@
 ^UNITTESTS
 ^storage/blockdevice/tests/UNITTESTS
 ^storage/kvstore/tests/UNITTESTS
-^drivers/tests/UNITTESTS
```
Just an observation: There seem to be no style changes needed in this commit. The style checking CI is passing. If there were files out of style in drivers/tests/UNITTESTS, this commit would have needed fixes to get those files back in style.

hazzlim replied:
Thanks Jaeden, noted. I did run astyle on them but they turned out to all be fine, luckily.

Patater reviewed (approved):
LGTM
